pkg/shared/utils/state_machine_parser/statemachine_builder.go

```diff
@@ -49,26 +49,24 @@ func (smb *stateMachineBuilder) Build() State {
 
 	tokenMap := make(map[string]State)
 
-	tokenMap[StartState] = NewStartState()
-	tokenMap[EndState] = NewEndState()
+	tokenMap[StartState] = newStartState()
+	tokenMap[EndState] = newEndState()
 
 	// build all the tokens
 	for _, t := range smb.grammar.Tokens {
-		b := NewStateBuilder(t.Name).
+		tokenMap[t.Name] = NewStateBuilder(t.Name).
 			Family(t.Family).
 			AcceptPattern(t.AcceptPattern).
-			OnNewToken(smb.onNewToken)
-
-		token := b.Build()
-		tokenMap[t.Name] = token
+			OnNewToken(smb.onNewToken).
+			Build()
 	}
 
 	// add all the transitions
 	for _, transition := range smb.grammar.Transitions {
 		token := tokenMap[transition.TokenName]
 		for _, targetTokenName := range transition.ValidTransitions {
 			targetToken := tokenMap[targetTokenName]
-			token.addNextState(targetToken)
+			token.(*state).addNextState(targetToken)
```

Comment:
Why the need to `.(*state).`?

Reply:
So it is safe to say that not all the implementation of the `State` interface will have the `addNextState` method?
I wonder what happens then with this cast.

Reply:
Nope. We will have only one implementation of the State interface, however, the State interface exposes only readonly methods.
There is no way to replace the internal representation of the State with another implementation.

Reply:
Okay. It make sense. To remove the temptation, I'd really remove that method from the *State* implementation and pass whatever that needs to be set to the State as a constructor argument i.e construction/instanciation only  argument. Because, this seems like what the code is doing here.